# Exploratory Data Analysis – Online Retail Dataset

## 🎯 Business Objective
The objective of this project is to analyze retail transaction data to understand customer purchasing behavior, revenue trends, and product demand patterns.

## 🏢 Business Context
This dataset contains transactional data from a UK-based online retail store. Each row represents a product purchased within a transaction.

The dataset includes:
- Transaction details
- Product information
- Customer details
- Purchase timestamps

## ❓ Business Questions
- Which country generates the highest revenue?
- What products sell the most?
- Are there seasonal sales patterns?
- Are there abnormal transactions or outliers?
- What factors influence revenue?

## 📊 Expected Impact
The analysis will help understand sales performance, customer behavior, and revenue drivers.


In [2]:
import pandas as pd
import numpy as np

In [4]:
df= pd.read_csv(r"C:\Users\niyas\python\EDA Project\E-Commerce Analysis - UK\data\raw\data.csv")
df.head()
    

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## Dataset Preview Observations

- Dataset successfully loaded.
- Dataset contains transaction-level retail data.
- Each row represents a product purchased in a transaction.


In [5]:
df.shape

(541909, 8)

## Dataset Shape

The dataset contains 541909 rows and 8 columns.


In [6]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

## Column Overview

The dataset contains the following columns:
- InvoiceNo
- StockCode
- Description
- Quantity
- InvoiceDate
- UnitPrice
- CustomerID
- Country


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [9]:
df["InvoiceDate"]= pd.to_datetime(df["InvoiceDate"])

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [11]:
df.head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

## Data Types Analysis

- Numerical columns: Quantity, UnitPrice
- Categorical columns: InvoiceNo, StockCode, Description, Country
- Date column: InvoiceDate
- CustomerID stored as numeric but represents customer identifier.


In [12]:
df.describe()

Quantity                    InvoiceDate      UnitPrice  \
count  541909.000000                         541909  541909.000000   
mean        9.552250  2011-07-04 13:34:57.156386048       4.611114   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 11:34:00       1.250000   
50%         3.000000            2011-07-19 17:17:00       2.080000   
75%        10.000000            2011-10-19 11:27:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       218.081158                            NaN      96.759853   

          CustomerID  
count  406829.000000  
mean    15287.690570  
min     12346.000000  
25%     13953.000000  
50%     15152.000000  
75%     16791.000000  
max     18287.000000  
std      1713.600303

## Statistical Summary Observations

- Quantity contains negative values indicating possible product returns.
- UnitPrice shows large variation suggesting potential outliers.
- Revenue-related fields show wide spread in values.


In [14]:
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

## Unique Value Analysis

- InvoiceNo and StockCode contain high number of unique values indicating identifier columns.
- Country contains limited unique values representing categorical variable.
- CustomerID represents unique customers.
- Description contains product names.


In [15]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [17]:
(df.isnull().sum()/len(df))*100

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64

## Missing Value Analysis

- CustomerID contains significant missing values.
- Description contains small number of missing values.
- Other columns have minimal or no missing data.

These missing values will be handled during data cleaning stage.


In [22]:
df.duplicated()

np.int64(541909)

In [19]:
df[df.duplicated()].head()

InvoiceNo StockCode                        Description  Quantity  \
517    536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527    536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537    536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539    536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555    536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517 2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
527 2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
537 2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
539 2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
555 2010-12-01 11:49:00       2.95     17920.0  United Kingdom

## Duplicate Record Analysis

- Dataset contains 540919 duplicate records.
- Duplicate transactions may affect revenue and sales analysis.
- Duplicate handling will be performed during data cleaning stage.


# Data Quality Issue Log

Based on initial data profiling, the following data quality issues were identified:

## 1. Missing Values
- CustomerID contains significant missing values.
- Description contains few missing values.

## 2. Incorrect Data Types
- InvoiceDate was initially stored as object and converted to datetime.
- CustomerID stored as numeric but represents identifier.

## 3. Negative Values
- Quantity contains negative values indicating possible product returns or cancellations.

## 4. Outliers
- UnitPrice shows large variation suggesting potential outliers.
- Quantity also contains extreme values.

## 5. Duplicate Records
- Dataset contains duplicate rows that may affect analysis.

## 6. High Cardinality Columns
- InvoiceNo and StockCode contain many unique values (identifier columns).

## Cleaning Strategy (To be implemented Day 2)
- Handle missing values
- Remove duplicates
- Treat outliers
- Standardize data types


In [23]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [24]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [25]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom


## Data Cleaning and Preprocessing

Goal:
- Handle missing values
- Remove duplicates
- Fix data quality issues
- Detect and treat outliers
- Prepare dataset for analysis


In [26]:
df.duplicated().sum()

np.int64(5268)

In [27]:
df.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Length: 541909, dtype: bool

In [28]:
df.drop_duplicates(inplace=True)

In [30]:
df.duplicated().sum()

np.int64(0)

Duplicate records removed to ensure datacon